### PR TITLE
BUG: Normalize address before comparison

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/dask/distributed/issues/2058

### CODE REVIEW

**Comment by TomAugspurger** [comment]
cc @mrocklin 

The test is good, but I'm not sure if this is the appropriate fix or not. My idea was to always normalize user-provided `address`es. `resolve_address` seemed to fit the bill.

If you like the fix, then I'll go through `worker.py` and `client.py` to always resolve the address on user-defined input.

**Comment by mrocklin** [comment]
I don't have strong thoughts about how this could backfire.  @pitrou is the expert here.  I'm inclined to go ahead with it if the test suite doesn't complain.

**Comment by TomAugspurger** [comment]
I notice that other places accepting an `address` also take a `resolve_address` keyword argument. e.g. `Scheduler.add_worker`. I'll plan on following that approach.

**Comment by mrocklin** [comment]
The windows failure is interesting.  I haven't seen it before and *may* be related?  I'm not sure.

File: distributed/tests/test_worker_client.py [hunk 1]
```diff
@@ -297,3 +297,17 @@ def f():
...
+    # https://github.com/dask/distributed/issues/2058
+    da = pytest.importorskip('dask.array')
+    c = yield Client('localhost:' + s.address.split(":")[-1], loop=s.loop,
+                     asynchronous=True)
+    with c:
+        X = c.persist(da.random.uniform(size=(100, 10), chunks=50))
+        yield wait(X)
+
+        fut = yield c.submit(lambda x: x.sum().compute(), X)
+        assert fut > 0
```
Comment (mrocklin) [inline distributed/tests/test_worker_client.py]: I recommend avoiding the use of the normal context manager in async tests.  If you were to use `async with c` that would work well, but we can't use that within tests that have to be within python 2 friendly files.  Instead I recommend using try/finally.  

Also, if you're not going to use the client created in the gen_cluster decorator then you probably want the following:

```python
@gen_cluster()
def test_foo(s, a, b):
    c = yield Client(...)

    try:
        ...
    finally:
        yield c.close()
```

There are also some examples of this in test_client.py

Comment (TomAugspurger) [reply distributed/tests/test_worker_client.py]: OK, thanks. I was just using the context manager so that it was cleaned up.

Comment (mrocklin) [reply distributed/tests/test_worker_client.py]: Understood, but the context manager is synchronous API.  There is no way for us to `yield c.close()` within the normal context mananger.  I think that it just starts the close coroutine, but doesn't wait for it.  You'll either need to use `async with` or explicitly call `c.close()` to be sure that it closes up.

**Comment by TomAugspurger** [comment]
Agreed that the windows test could be related. I'll see if it fails this time around.

**Comment by TomAugspurger** [comment]
Windows passed this time.

The Python 2 travis run failed: https://travis-ci.org/dask/distributed/jobs/394751774#L1637

File: distributed/tests/test_worker_client.py [hunk 2]
```diff
@@ -297,3 +297,19 @@ def f():
...
+
+@gen_cluster()
+def test_submit_different_names(s, a, b):
+    # https://github.com/dask/distributed/issues/2058
+    da = pytest.importorskip('dask.array')
+    c = yield Client('localhost:' + s.address.split(":")[-1], loop=s.loop,
+                     asynchronous=True)
+    try:
+        X = c.persist(da.random.uniform(size=(100, 10), chunks=50))
+        yield wait(X)
```
Comment (mrocklin) [inline distributed/tests/test_worker_client.py]: Fine for now, but in the future I recommend `foo.persist()` over `c.persist(foo)`

Comment (mrocklin) [reply distributed/tests/test_worker_client.py]: Just for aesthetic reasons really

**Pull request merged**